File: Binary Search Tree Iterator.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class BSTIterator {

    public BSTIterator(TreeNode root) {
        TreeNode node = root;
        while (node != null)
        {
            m_nodes.push(node);
            node = node.left;
        }
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return !m_nodes.isEmpty();
    }

    /** @return the next smallest number */
    public int next() {
        if (m_nodes.isEmpty())
        {
            //throw exception
            return 0;
        }
        TreeNode node = m_nodes.pop();
        int val = node.val;
        if (node.right != null)
        {
            node = node.right;
            m_nodes.push(node);
            while (node.left != null)
            {
                node = node.left;
                m_nodes.push(node);
            }
        }
        return val;
    }
    
    private Deque<TreeNode> m_nodes = new ArrayDeque<TreeNode>();
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: Binary Search Tree Iterator.py
# Definition for a  binary tree node
# class TreeNode:
#	 def __init__(self, x):
#		 self.val = x
#		 self.left = None
#		 self.right = None

class BSTIterator:
	# @param root, a binary search tree's root node
	def __init__(self, root):
		node = root
		self.nodes = []
		while (node != None):
			self.nodes.append(node)
			node = node.left

	# @return a boolean, whether we have a next smallest number
	def hasNext(self):
		return (len(self.nodes) != 0)

	# @return an integer, the next smallest number
	def next(self):
		last = len(self.nodes)-1
		node = self.nodes.pop(last)
		val = node.val
		if (node.right):
			node = node.right
			self.nodes.append(node)
			while (node.left):
				node = node.left
				self.nodes.append(node)
		return val
		

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())